sway-lib-std/src/storage.sw

```diff
@@ -681,6 +681,262 @@ impl<V> StorageVec<V> {
+            _ => get::<V>(sha256((0, __get_storage_key()))),
+        }
+    }
+
+    /// Returns the last element of the vector, or `None` if it is empty.
+    ///
+    /// ### Storage Access
+    ///
+    /// * Reads: `2`
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     vec: StorageVec<u64> = StorageVec {},
+    ///     vec2: StorageVec<u64> = StorageVec {},
+    /// }
+    ///
+    /// fn foo() {
+    ///     storage.vec.push(5);
+    ///     storage.vec.push(10);
+    ///
+    ///     assert(10 == storage.vec.last().unwrap());
+    ///     assert(storage.vec2.last().is_none());
+    /// }
+    /// ```
+    #[storage(read)]
+    pub fn last(self) -> Option<V> {
+        match get::<u64>(__get_storage_key()).unwrap_or(0) {
+            0 => Option::None,
+            len => get::<V>(sha256((len - 1, __get_storage_key()))),
+        }
+    }
+
+    /// Reverses the order of elements in the vector, in place.
+    ///
+    /// ### Storage Access
+    ///
+    /// * Reads: `1 + (2 * (self.len() / 2))`
+    /// * Writes: `2 * (self.len() / 2)`
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     vec: StorageVec<u64> = StorageVec {},
+    /// }
+    ///
+    /// fn foo() {
+    ///     storage.vec.push(5);
+    ///     storage.vec.push(10);
+    ///     storage.vec.push(15);
+    ///     storage.vec.reverse();
+    ///
+    ///     assert(15 == storage.vec.get(0).unwrap());
+    ///     assert(10 == storage.vec.get(1).unwrap());
+    ///     assert(5 == storage.vec.get(2).unwrap());
+    /// }
+    /// ```
+    #[storage(read, write)]
+    pub fn reverse(self) {
+        let len = get::<u64>(__get_storage_key()).unwrap_or(0);
+
+        if len < 2 {
+            return;
+        }
+
+        let mid = len / 2;
+        let mut i = 0;
+        while i < mid {
+            let element1_key = sha256((i, __get_storage_key()));
+            let element2_key = sha256((len - i - 1, __get_storage_key()));
+
+            let element1_value: V = get::<V>(element1_key).unwrap();
+            store::<V>(element1_key, get::<V>(element2_key).unwrap());
+            store::<V>(element2_key, element1_value);
+
+            i += 1;
+        }
+    }
+
+    /// Fills `self` with elements by cloning `value`.
+    ///
+    /// ### Arguments
+    ///
+    /// * value - Value to copy to each element of the vector.
+    ///
+    /// ### Storage Access
+    ///
+    /// * Reads: `1`
+    /// * Writes: `self.len()`
+    ///
+    /// ### Examples
+    ///
+    /// ```sway
+    /// storage {
+    ///     vec: StorageVec<u64> = StorageVec {},
+    /// }
+    ///
+    /// fn foo() {
+    ///     storage.vec.push(5);
+    ///     storage.vec.push(10);
+    ///     storage.vec.push(15);
+    ///     storage.vec.fill(20);
+    ///
+    ///     assert(20 == storage.vec.get(0).unwrap());
+    ///     assert(20 == storage.vec.get(1).unwrap());
+    ///     assert(20 == storage.vec.get(2).unwrap());
+    /// }
```

Comment:
Wondering about the `push` here.

On one hand you're clearly indicating that different values are added and thus the assertions state that they have all been overwritten.

On the other hand, we should encourage users to create vectors from a known capacity to avoid reallocating. If this approached is used then we still indicate that all values are as we expect; however, a user may not be as familiar with the capacity approach compared to pushing.

I think this should be altered to use the capacity as overall it's a better standard and more educational.

Reply:
Storage vectors don't have capacity, so there is no reallocation. Though this would be a good idea for `Vec` examples.

Reply:
naruhodo naruhodo